File: spine-ts/core/src/attachments/RegionAttachment.ts
/******************************************************************************
 * Spine Runtimes License Agreement
 * Last updated January 1, 2020. Replaces all prior versions.
 *
 * Copyright (c) 2013-2020, Esoteric Software LLC
 *
 * Integration of the Spine Runtimes into software or otherwise creating
 * derivative works of the Spine Runtimes is permitted under the terms and
 * conditions of Section 2 of the Spine Editor License Agreement:
 * http://esotericsoftware.com/spine-editor-license
 *
 * Otherwise, it is permitted to integrate the Spine Runtimes into software
 * or otherwise create derivative works of the Spine Runtimes (collectively,
 * "Products"), provided that each user of the Products must obtain their own
 * Spine Editor license and redistribution of the Products in any form must
 * include this license and copyright notice.
 *
 * THE SPINE RUNTIMES ARE PROVIDED BY ESOTERIC SOFTWARE LLC "AS IS" AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL ESOTERIC SOFTWARE LLC BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES,
 * BUSINESS INTERRUPTION, OR LOSS OF USE, DATA, OR PROFITS) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THE SPINE RUNTIMES, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *****************************************************************************/

module spine {

	/** An attachment that displays a textured quadrilateral.
	 *
	 * See [Region attachments](http://esotericsoftware.com/spine-regions) in the Spine User Guide. */
	export class RegionAttachment extends Attachment {
		static OX1 = 0;
		static OY1 = 1;
		static OX2 = 2;
		static OY2 = 3;
		static OX3 = 4;
		static OY3 = 5;
		static OX4 = 6;
		static OY4 = 7;

		static X1 = 0;
		static Y1 = 1;
		static C1R = 2;
		static C1G = 3;
		static C1B = 4;
		static C1A = 5;
		static U1 = 6;
		static V1 = 7;

		static X2 = 8;
		static Y2 = 9;
		static C2R = 10;
		static C2G = 11;
		static C2B = 12;
		static C2A = 13;
		static U2 = 14;
		static V2 = 15;

		static X3 = 16;
		static Y3 = 17;
		static C3R = 18;
		static C3G = 19;
		static C3B = 20;
		static C3A = 21;
		static U3 = 22;
		static V3 = 23;

		static X4 = 24;
		static Y4 = 25;
		static C4R = 26;
		static C4G = 27;
		static C4B = 28;
		static C4A = 29;
		static U4 = 30;
		static V4 = 31;

		/** The local x translation. */
		x = 0;

		/** The local y translation. */
		y = 0;

		/** The local scaleX. */
		scaleX = 1;

		/** The local scaleY. */
		scaleY = 1;

		/** The local rotation. */
		rotation = 0;

		/** The width of the region attachment in Spine. */
		width = 0;

		/** The height of the region attachment in Spine. */
		height = 0;

		/** The color to tint the region attachment. */
		color = new Color(1, 1, 1, 1);

		/** The name of the texture region for this attachment. */
		path: string;

		rendererObject: any;
		region: TextureRegion;

		/** For each of the 4 vertices, a pair of <code>x,y</code> values that is the local position of the vertex.
		 *
		 * See {@link #updateOffset()}. */
		offset = Utils.newFloatArray(8);


		uvs = Utils.newFloatArray(8);

		tempColor = new Color(1, 1, 1, 1);

		constructor (name:string) {
			super(name);
		}

		/** Calculates the {@link #offset} using the region settings. Must be called after changing region settings. */
		updateOffset () : void {
			let regionScaleX = this.width / this.region.originalWidth * this.scaleX;
			let regionScaleY = this.height / this.region.originalHeight * this.scaleY;
			let localX = -this.width / 2 * this.scaleX + this.region.offsetX * regionScaleX;
			let localY = -this.height / 2 * this.scaleY + this.region.offsetY * regionScaleY;
			let localX2 = localX + this.region.width * regionScaleX;
			let localY2 = localY + this.region.height * regionScaleY;
			let radians = this.rotation * Math.PI / 180;
			let cos = Math.cos(radians);
			let sin = Math.sin(radians);
			let localXCos = localX * cos + this.x;
			let localXSin = localX * sin;
			let localYCos = localY * cos + this.y;
			let localYSin = localY * sin;
			let localX2Cos = localX2 * cos + this.x;
			let localX2Sin = localX2 * sin;
			let localY2Cos = localY2 * cos + this.y;
			let localY2Sin = localY2 * sin;
			let offset = this.offset;
			offset[RegionAttachment.OX1] = localXCos - localYSin;
			offset[RegionAttachment.OY1] = localYCos + localXSin;
			offset[RegionAttachment.OX2] = localXCos - localY2Sin;
			offset[RegionAttachment.OY2] = localY2Cos + localXSin;
			offset[RegionAttachment.OX3] = localX2Cos - localY2Sin;
			offset[RegionAttachment.OY3] = localY2Cos + localX2Sin;
			offset[RegionAttachment.OX4] = localX2Cos - localYSin;
			offset[RegionAttachment.OY4] = localYCos + localX2Sin;
		}

		setRegion (region: TextureRegion) : void {
			this.region = region;
			let uvs = this.uvs;
			if (region.rotate) {
				uvs[2] = region.u;
				uvs[3] = region.v2;
				uvs[4] = region.u;
				uvs[5] = region.v;
				uvs[6] = region.u2;
				uvs[7] = region.v;
				uvs[0] = region.u2;
				uvs[1] = region.v2;
			} else {
				uvs[0] = region.u;
				uvs[1] = region.v2;
				uvs[2] = region.u;
				uvs[3] = region.v;
				uvs[4] = region.u2;
				uvs[5] = region.v;
				uvs[6] = region.u2;
				uvs[7] = region.v2;
			}
		}

		/** Transforms the attachment's four vertices to world coordinates.
		 *
		 * See [World transforms](http://esotericsoftware.com/spine-runtime-skeletons#World-transforms) in the Spine
		 * Runtimes Guide.
		 * @param worldVertices The output world vertices. Must have a length >= `offset` + 8.
		 * @param offset The `worldVertices` index to begin writing values.
		 * @param stride The number of `worldVertices` entries between the value pairs written. */
		computeWorldVertices (bone: Bone, worldVertices: ArrayLike<number>, offset: number, stride: number) {
			let vertexOffset = this.offset;
			let x = bone.worldX, y = bone.worldY;
			let a = bone.a, b = bone.b, c = bone.c, d = bone.d;
			let offsetX = 0, offsetY = 0;

			offsetX = vertexOffset[RegionAttachment.OX1];
			offsetY = vertexOffset[RegionAttachment.OY1];
			worldVertices[offset] = offsetX * a + offsetY * b + x; // br
			worldVertices[offset + 1] = offsetX * c + offsetY * d + y;
			offset += stride;

			offsetX = vertexOffset[RegionAttachment.OX2];
			offsetY = vertexOffset[RegionAttachment.OY2];
			worldVertices[offset] = offsetX * a + offsetY * b + x; // bl
			worldVertices[offset + 1] = offsetX * c + offsetY * d + y;
			offset += stride;

			offsetX = vertexOffset[RegionAttachment.OX3];
			offsetY = vertexOffset[RegionAttachment.OY3];
			worldVertices[offset] = offsetX * a + offsetY * b + x; // ul
			worldVertices[offset + 1] = offsetX * c + offsetY * d + y;
			offset += stride;

			offsetX = vertexOffset[RegionAttachment.OX4];
			offsetY = vertexOffset[RegionAttachment.OY4];
			worldVertices[offset] = offsetX * a + offsetY * b + x; // ur
			worldVertices[offset + 1] = offsetX * c + offsetY * d + y;
		}

		copy (): Attachment {
			let copy = new RegionAttachment(this.name);
			copy.region = this.region;
			copy.rendererObject = this.rendererObject;
			copy.path = this.path;
			copy.x = this.x;
			copy.y = this.y;
			copy.scaleX = this.scaleX;
			copy.scaleY = this.scaleY;
			copy.rotation = this.rotation;
			copy.width = this.width;
			copy.height = this.height;
			Utils.arrayCopy(this.uvs, 0, copy.uvs, 0, 8);
			Utils.arrayCopy(this.offset, 0, copy.offset, 0, 8);
			copy.color.setFromColor(this.color);
			return copy;
		}
	}
}

File: spine-lua/spine-lua/attachments/RegionAttachment.lua
-------------------------------------------------------------------------------
-- Spine Runtimes License Agreement
-- Last updated January 1, 2020. Replaces all prior versions.
--
-- Copyright (c) 2013-2020, Esoteric Software LLC
--
-- Integration of the Spine Runtimes into software or otherwise creating
-- derivative works of the Spine Runtimes is permitted under the terms and
-- conditions of Section 2 of the Spine Editor License Agreement:
-- http://esotericsoftware.com/spine-editor-license
--
-- Otherwise, it is permitted to integrate the Spine Runtimes into software
-- or otherwise create derivative works of the Spine Runtimes (collectively,
-- "Products"), provided that each user of the Products must obtain their own
-- Spine Editor license and redistribution of the Products in any form must
-- include this license and copyright notice.
--
-- THE SPINE RUNTIMES ARE PROVIDED BY ESOTERIC SOFTWARE LLC "AS IS" AND ANY
-- EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
-- WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
-- DISCLAIMED. IN NO EVENT SHALL ESOTERIC SOFTWARE LLC BE LIABLE FOR ANY
-- DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
-- (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES,
-- BUSINESS INTERRUPTION, OR LOSS OF USE, DATA, OR PROFITS) HOWEVER CAUSED AND
-- ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
-- (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
-- THE SPINE RUNTIMES, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-------------------------------------------------------------------------------

local setmetatable = setmetatable
local math_pi = math.pi
local math_sin = math.sin
local math_cos = math.cos

local AttachmentType = require "spine-lua.attachments.AttachmentType"
local Attachment = require "spine-lua.attachments.Attachment"
local Color = require "spine-lua.Color"
local Utils = require "spine-lua.utils"

local OX1 = 1
local OY1 = 2
local OX2 = 3
local OY2 = 4
local OX3 = 5
local OY3 = 6
local OX4 = 7
local OY4 = 8

local X1 = 1
local Y1 = 2
local U1 = 3
local V1 = 4
local C1R = 5
local C1G = 6
local C1B = 7
local C1A = 8

local X2 = 9
local Y2 = 10
local U2 = 11
local V2 = 12
local C2R = 13
local C2G = 14
local C2B = 15
local C2A = 16

local X3 = 17
local Y3 = 18
local U3 = 19
local V3 = 20
local C3R = 21
local C3G = 22
local C3B = 23
local C3A = 24

local X4 = 25
local Y4 = 26
local U4 = 27
local V4 = 28
local C4R = 29
local C4G = 30
local C4B = 31
local C4A = 32

local RegionAttachment = {}
RegionAttachment.__index = RegionAttachment
setmetatable(RegionAttachment, { __index = Attachment })

RegionAttachment.OX1 = 1
RegionAttachment.OY1 = 2
RegionAttachment.OX2 = 3
RegionAttachment.OY2 = 4
RegionAttachment.OX3 = 5
RegionAttachment.OY3 = 6
RegionAttachment.OX4 = 7
RegionAttachment.OY4 = 8

RegionAttachment.X1 = 1
RegionAttachment.Y1 = 2
RegionAttachment.U1 = 3
RegionAttachment.V1 = 4
RegionAttachment.C1R = 5
RegionAttachment.C1G = 6
RegionAttachment.C1B = 7
RegionAttachment.C1A = 8

RegionAttachment.X2 = 9
RegionAttachment.Y2 = 10
RegionAttachment.U2 = 11
RegionAttachment.V2 = 12
RegionAttachment.C2R = 13
RegionAttachment.C2G = 14
RegionAttachment.C2B = 15
RegionAttachment.C2A = 16

RegionAttachment.X3 = 17
RegionAttachment.Y3 = 18
RegionAttachment.U3 = 19
RegionAttachment.V3 = 20
RegionAttachment.C3R = 21
RegionAttachment.C3G = 22
RegionAttachment.C3B = 23
RegionAttachment.C3A = 24

RegionAttachment.X4 = 25
RegionAttachment.Y4 = 26
RegionAttachment.U4 = 27
RegionAttachment.V4 = 28
RegionAttachment.C4R = 29
RegionAttachment.C4G = 30
RegionAttachment.C4B = 31
RegionAttachment.C4A = 32

function RegionAttachment.new (name)
	if not name then error("name cannot be nil", 2) end

	local self = Attachment.new(name, AttachmentType.region)
	self.x = 0
	self.y = 0
	self.scaleX = 1
	self.scaleY = 1
	self.rotation = 0
	self.width = 0
	self.height = 0
	self.color = Color.newWith(1, 1, 1, 1)
	self.path = nil
	self.rendererObject = nil
	self.region = nil
	self.offset = Utils.newNumberArray(8)
	self.uvs = Utils.newNumberArray(8)
	self.tempColor = Color.newWith(1, 1, 1, 1)
	setmetatable(self, RegionAttachment)

	return self
end

function RegionAttachment:updateOffset ()
	local regionScaleX = self.width / self.region.originalWidth * self.scaleX
	local regionScaleY = self.height / self.region.originalHeight * self.scaleY
	local localX = -self.width / 2 * self.scaleX + self.region.offsetX * regionScaleX
	local localY = -self.height / 2 * self.scaleY + self.region.offsetY * regionScaleY
	local localX2 = localX + self.region.width * regionScaleX
	local localY2 = localY + self.region.height * regionScaleY
	local radians = self.rotation * math_pi / 180
	local cos = math_cos(radians)
	local sin = math_sin(radians)
	local localXCos = localX * cos + self.x
	local localXSin = localX * sin
	local localYCos = localY * cos + self.y
	local localYSin = localY * sin
	local localX2Cos = localX2 * cos + self.x
	local localX2Sin = localX2 * sin
	local localY2Cos = localY2 * cos + self.y
	local localY2Sin = localY2 * sin
	local offset = self.offset
	offset[OX1] = localXCos - localYSin
	offset[OY1] = localYCos + localXSin
	offset[OX2] = localXCos - localY2Sin
	offset[OY2] = localY2Cos + localXSin
	offset[OX3] = localX2Cos - localY2Sin
	offset[OY3] = localY2Cos + localX2Sin
	offset[OX4] = localX2Cos - localYSin
	offset[OY4] = localYCos + localX2Sin
end

function RegionAttachment:setRegion (region)
	local uvs = self.uvs
	if region.rotate then
		uvs[5] = region.u
		uvs[6] = region.v2
		uvs[7] = region.u
		uvs[8] = region.v
		uvs[1] = region.u2
		uvs[2] = region.v
		uvs[3] = region.u2
		uvs[4] = region.v2
	else
		uvs[3] = region.u
		uvs[4] = region.v2
		uvs[5] = region.u
		uvs[6] = region.v
		uvs[7] = region.u2
		uvs[8] = region.v
		uvs[1] = region.u2
		uvs[2] = region.v2
	end
end

function RegionAttachment:computeWorldVertices (bone, worldVertices, offset, stride)
	offset = offset + 1
	local vertexOffset = self.offset
	local x = bone.worldX
	local y = bone.worldY
	local a = bone.a
	local b = bone.b
	local c = bone.c
	local d = bone.d
	local offsetX = 0
	local offsetY = 0

	offsetX = vertexOffset[7]
	offsetY = vertexOffset[8]
	worldVertices[offset] = offsetX * a + offsetY * b + x -- br
	worldVertices[offset + 1] = offsetX * c + offsetY * d + y
	offset = offset + stride

	offsetX = vertexOffset[1]
	offsetY = vertexOffset[2]
	worldVertices[offset] = offsetX * a + offsetY * b + x -- bl
	worldVertices[offset + 1] = offsetX * c + offsetY * d + y
	offset = offset + stride

	offsetX = vertexOffset[3]
	offsetY = vertexOffset[4]
	worldVertices[offset] = offsetX * a + offsetY * b + x -- ul
	worldVertices[offset + 1] = offsetX * c + offsetY * d + y
	offset = offset + stride

	offsetX = vertexOffset[5]
	offsetY = vertexOffset[6]
	worldVertices[offset] = offsetX * a + offsetY * b + x -- ur
	worldVertices[offset + 1] = offsetX * c + offsetY * d + y
end

function RegionAttachment:copy ()
	local copy = RegionAttachment.new(self.name)
	copy.x = self.x
	copy.y = self.y
	copy.scaleX = self.scaleX
	copy.scaleY = self.scaleY
	copy.rotation = self.rotation
	copy.width = self.width
	copy.height = self.height
	copy.color:setFrom(self.color)
	copy.path = self.path
	copy.rendererObject = self.rendererObject
	copy.region = self.region
	copy.offset = Utils.copy(self.offset)
	copy.uvs = Utils.copy(self.uvs)
	copy.tempColor:setFrom(self.tempColor)
	return copy
end

return RegionAttachment
